File: daybreak/src/daybreak/core/theme_model.py
from typing import Dict

THEME_MODES = {"light", "dark"}

TOKEN_KEYS = (
    "bg",
    "fg",
    "cursor",
    "muted",
    "primary",
    "success",
    "warning",
    "error",
    "info",
    "surface_1",
    "surface_2",
    "surface_3",
)


def normalize_mode(mode: str) -> str:
    normalized = (mode or "").strip().lower()
    if normalized not in THEME_MODES:
        raise ValueError(f"Unsupported mode '{mode}'. Expected one of: {sorted(THEME_MODES)}")
    return normalized


def is_hex_color(value: str) -> bool:
    if not isinstance(value, str):
        return False
    if len(value) != 7 or not value.startswith("#"):
        return False
    try:
        int(value[1:], 16)
    except ValueError:
        return False
    return True


def validate_tokens(tokens: Dict[str, str]) -> None:
    missing = [key for key in TOKEN_KEYS if key not in tokens]
    if missing:
        raise ValueError(f"Theme tokens missing required keys: {missing}")

    invalid = [key for key, value in tokens.items() if key in TOKEN_KEYS and not is_hex_color(value)]
    if invalid:
        raise ValueError(f"Theme tokens contain invalid hex colors for keys: {invalid}")

File: src/core/themeModel.ts
export const THEME_MODES: Set<string> = new Set(['light', 'dark']);

export const TOKEN_KEYS: readonly string[] = [
  'bg',
  'fg',
  'cursor',
  'muted',
  'primary',
  'success',
  'warning',
  'error',
  'info',
  'surface_1',
  'surface_2',
  'surface_3',
] as const;

export const ACCENT_KEYS: readonly string[] = [
  'accent_primary',
  'accent_secondary',
  'accent_success',
  'accent_warning',
  'accent_error',
  'accent_selection',
] as const;

export function normalizeMode(mode: string): string {
  const normalized = (mode || '').trim().toLowerCase();
  if (!THEME_MODES.has(normalized)) {
    throw new Error(`Unsupported mode '${mode}'. Expected one of: ${[...THEME_MODES].sort().join(', ')}`);
  }
  return normalized;
}

export function isHexColor(value: string): boolean {
  if (typeof value !== 'string') return false;
  if (value.length !== 7 || !value.startsWith('#')) return false;
  try {
    parseInt(value.slice(1), 16);
    return true;
  } catch {
    return false;
  }
}

export function validateTokens(tokens: Record<string, string>): void {
  const missing = TOKEN_KEYS.filter(key => !(key in tokens));
  if (missing.length > 0) {
    throw new Error(`Theme tokens missing required keys: ${JSON.stringify(missing)}`);
  }
  const invalid = Object.entries(tokens)
    .filter(([key, value]) => TOKEN_KEYS.includes(key) && !isHexColor(value))
    .map(([key]) => key);
  if (invalid.length > 0) {
    throw new Error(`Theme tokens contain invalid hex colors for keys: ${JSON.stringify(invalid)}`);
  }
}
